File: code-analyzer/examples/sample.ts
/**
 * Sample TypeScript module for analysis
 */

interface UserData {
  id: string;
  name: string;
  email: string;
  role: string;
}

/**
 * API Service - handles too much
 */
class APIService {
  private baseURL: string = "http://api.example.com";

  /**
   * Fetch user data - too many parameters
   */
  async fetchUserData(
    userId: string,
    includeProfile: boolean,
    includeSettings: boolean,
    includePosts: boolean,
    includeFollowers: boolean,
    includeFollowing: boolean,
    format: string
  ): Promise<UserData> {
    const url = `${this.baseURL}/users/${userId}`;
    const response = await fetch(url);
    return response.json();
  }

  /**
   * Callback hell - too many nested callbacks
   */
  loadUserWithCallback(userId: string, callback: Function) {
    this.fetchData(userId, (err: any, user: any) => {
      if (err) {
        callback(err);
      } else {
        this.fetchPermissions(user.id, (err2: any, perms: any) => {
          if (err2) {
            callback(err2);
          } else {
            this.fetchSettings(user.id, (err3: any, settings: any) => {
              callback(null, { user, perms, settings });
            });
          }
        });
      }
    });
  }

  private fetchData(id: string, callback: Function): void {
    // Implementation
  }

  private fetchPermissions(id: string, callback: Function): void {
    // Implementation
  }

  private fetchSettings(id: string, callback: Function): void {
    // Implementation
  }

  /**
   * Complex conditional logic
   */
  validateUserAccess(user: any): boolean {
    if (user !== null && user !== undefined) {
      if (user.isActive === true) {
        if (user.role === "admin" || user.role === "moderator") {
          if (user.permissions && user.permissions.length > 0) {
            if (user.permissions.includes("write")) {
              if (user.lastLoginDate) {
                return true;
              }
            }
          }
        }
      }
    }
    return false;
  }

  /**
   * Magic numbers throughout
   */
  calculateDelay(retryCount: number): number {
    if (retryCount === 0) {
      return 1000;
    } else if (retryCount === 1) {
      return 2000;
    } else if (retryCount === 2) {
      return 4000;
    } else if (retryCount === 3) {
      return 8000;
    } else if (retryCount === 4) {
      return 15000;
    }
    return 30000;
  }

  /**
   * Duplicate error handling
   */
  handleError1(error: any): void {
    if (error === null) {
      console.log("No error");
      return;
    }
    if (error === undefined) {
      console.log("Undefined error");
      return;
    }
    console.error("Error occurred:", error);
  }

  handleError2(error: any): void {
    if (error === null) {
      console.log("No error");
      return;
    }
    if (error === undefined) {
      console.log("Undefined error");
      return;
    }
    console.error("Error occurred:", error);
  }
}

/**
 * Utility functions with poor naming
 */
function processData(data: any): any {
  return data.map((item: any) => item * 2);
}

function validateData(x: any): boolean {
  return x !== null && x !== undefined && x.length > 0;
}

function transformData(obj: any): any {
  const result: any = {};
  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      result[key] = obj[key];
    }
  }
  return result;
}

File: code-analyzer/examples/sample.py
"""Sample Python module for analysis"""

import sys
import os
from typing import List, Dict, Optional
from dataclasses import dataclass


@dataclass
class User:
    """User information"""
    name: str
    email: str
    age: int
    active: bool = True


class UserManager:
    """Manages user operations - does too much!"""
    
    def __init__(self):
        self.users = []
        self.db_connection = None
    
    def load_users_from_file(self, file_path: str):
        """Load users from file"""
        with open(file_path, 'r') as f:
            data = f.read()
        return data
    
    def save_users_to_database(self, user_data: List[User]):
        """Save users to database"""
        for user in user_data:
            if user.active:
                self.db_connection.insert(user)
    
    def validate_user_email(self, email: str) -> bool:
        """Validate email format"""
        return '@' in email and '.' in email
    
    def process_and_store_users(self, file_path: str, users: List[User]):
        """Process users - way too long!"""
        # Load from file
        raw_data = self.load_users_from_file(file_path)
        
        # Process each user
        for user in users:
            # Validate
            if not self.validate_user_email(user.email):
                print(f"Invalid email: {user.email}")
                continue
            
            # Transform
            user.name = user.name.upper()
            
            # Store
            self.save_users_to_database([user])
    
    def handle_user_request(self, request_type: str, user_id: str, user_name: str, 
                           user_email: str, user_age: int, is_active: bool, 
                           department: str, role: str, permissions: List[str]) -> str:
        """Handle user request - too many parameters!"""
        if request_type == "create":
            user = User(name=user_name, email=user_email, age=user_age, active=is_active)
            self.users.append(user)
            return f"Created user: {user_name}"
        elif request_type == "delete":
            self.users = [u for u in self.users if u.name != user_name]
            return f"Deleted user: {user_name}"
        return "Unknown request"


def duplicate_validation(data):
    """Validation function with duplicate code"""
    if data is None:
        return False
    if len(data) == 0:
        return False
    if not isinstance(data, list):
        return False
    return True


def another_validation(items):
    """Another validation - repeated pattern"""
    if items is None:
        return False
    if len(items) == 0:
        return False
    if not isinstance(items, list):
        return False
    return True


# Magic numbers everywhere!
def calculate_discount(price: float) -> float:
    """Calculate discount - magic numbers!"""
    if price > 100:
        return price * 0.9
    elif price > 50:
        return price * 0.95
    elif price > 25:
        return price * 0.98
    return price


def deeply_nested_function(a, b, c, d):
    """Deeply nested conditional logic"""
    if a:
        if b:
            if c:
                if d:
                    if a and b and c:
                        return "deeply nested"
    return None


# Bad exception handling
def risky_operation():
    """Silent exception handling"""
    try:
        result = 10 / 0
    except:
        pass
    return result
